### PR TITLE
[#112] MaxDepth & MinSample

### PR DESCRIPTION
**Pull request by Yatish0833**
Adding command line argument for max depth and min node size

### CODE REVIEW

File: src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala [hunk 1]
```diff
@@ -104,7 +104,16 @@ class ImportanceCmd extends ArgsApp with SparkApp
 
   @Option(name="-sr", required=false, usage="Random seed to use (def=<random>)", aliases=Array("--seed"))
   val randomSeed: Long = defRng.nextLong
-   
+
```
Comment (piotrszul) [inline src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala]: The current convention is to start all the random forest related options with `-rxxx` and `--rf-xxx` , e.g `-rmt` and `--rf-mtry`.
So there should be renamed to `-rmd` and `--rf-max-depth` etc.
Also, please move them up above `-sr` so that they are together with other random forest options.

(Nice to have/consider):
Refactoring all of the random forest options out of `ImportanceCmd` to `args.RandomForestArgs` in the same way as for instance `ImportanceOutputArgs`

File: src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala [hunk 2]
```diff
@@ -124,6 +133,8 @@ class ImportanceCmd extends ArgsApp with SparkApp
   
     val inputData = DefTreeRepresentationFactory.createRepresentation(featureSource.features.zipWithIndex()).cache() 
     val totalVariables = inputData.count()
+    echo(s"Input data: ${inputData}")
```
Comment (piotrszul) [inline src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala]: Why is this needed?

File: src/main/scala/au/csiro/variantspark/algo/RandomForest.scala [hunk 3]
```diff
@@ -220,7 +227,7 @@ class RandomForest(params:RandomForestParams=RandomForestParams()
     val oobAggregator = if (actualParams.oob) Option(new VotingAggregator(nLabels,nSamples)) else None
 
     val builder = modelBuilderFactory(DecisionTreeParams(seed = rng.nextLong, randomizeEquality = actualParams
-      .randomizeEquality))
+      .randomizeEquality, maxDepth = actualParams.maxDepth, minNodeSize = actualParams.minNodeSize))
```
Comment (piotrszul) [inline src/main/scala/au/csiro/variantspark/algo/RandomForest.scala]: Nice to have:
Refactoring: Perhaps it's worth adding a method to `RandomForestParam` to that produces `DecisionTreeParams` and only takes the seed.
E.e. `RandomForestParam.toDecisionTreeParams(seed:Int)`

**Review by piotrszul — COMMENTED** [review]
In general looks good!
I added some specific comments inside the code.
The only general issues is that you have introduced some formatting changes (lot's of new lines and one instance of a space at EOL) and I am not sure what's the motivation for these was. But they make the review harder, so I would like to discuss that (and possibly revert them back).

Do you have any ideas on how to test these new options.

**Comment by Yatish0833** [comment]
> In general looks good!
> I added some specific comments inside the code.
> The only general issues is that you have introduced some formatting changes (lot's of new lines and one instance of a space at EOL) and I am not sure what's the motivation for these was. But they make the review harder, so I would like to discuss that (and possibly revert them back).
> 
> Do you have any ideas on how to test these new options.

I understand reviewing new lines is actually a waste of time. I tried to remove all such instances but apparently, I missed quite a few. They were introduced mainly because I wrote a lot of print statements to understand the code.
I did test these new options locally but maybe I can write some meaningful unit test to make sure they are doing what they are supposed to do. 
I will look at the specific comments, fix them and update the build as well and make a new pull request.

File: src/main/scala/au/csiro/variantspark/algo/RandomForest.scala (hunk 3 above)
Comment (Yatish0833) [reply src/main/scala/au/csiro/variantspark/algo/RandomForest.scala]: Do you mean something like this?
def toDecisionTreeParams(seed:Int): DecisionTreeParams = { 
       DecisionTreeParams(seed = seed, randomizeEquality = actualParams.randomizeEquality, maxDepth = 
            actualParams.maxDepth, minNodeSize = actualParams.minNodeSize)
}

Such that the above call can be changed to:
val builder = modelBuilderFactory(RandomForestParam.toDecisionTreeParams(seed = rng.nextLong))

Is my understanding correct about this comment?

File: src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala (hunk 2 above)
Comment (Yatish0833) [reply src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala]: I must have missed this. Removing it now.

File: src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala (hunk 1 above)
Comment (Yatish0833) [reply src/main/scala/au/csiro/variantspark/cli/ImportanceCmd.scala]: Updated the code to reflect both of these changes.

File: src/main/scala/au/csiro/variantspark/algo/RandomForest.scala (hunk 3 above)
Comment (piotrszul) [reply src/main/scala/au/csiro/variantspark/algo/RandomForest.scala]: Yes (mostly). This should be an instance method (not a class method)
In RandomForestParam

`def toDecisionTreeParams(seed:Int): DecisionTreeParams = {
DecisionTreeParams(seed = seed, randomizeEquality = randomizeEquality, maxDepth =
maxDepth, minNodeSize = minNodeSize) // or minNodeSize = this.minNodeSize
}`

Usage (no need to pass arguments by name)

`val builder = modelBuilderFactory(actualParams.toDecisionTreeParams(rng.nextLong))`

Comment (Yatish0833) [reply src/main/scala/au/csiro/variantspark/algo/RandomForest.scala]: Updated the same

**Comment by piotrszul** [comment]
Merged. Good job!
